### PR TITLE
Multiplayer Game Transition and Endless Mode

### PR DESCRIPTION
**Pull request by brianyenna**
This PR implements the following features:
1. Multiplayer Game Transition
- Game can now transition to MultiplayerGameScene.

2. Endless Mode
- Endless Mode has been implemented and can be seen in the Multiplayer game. However, Endless Mode has not been ended for the single player game.

### CODE REVIEW

File: GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ This class handles how to add new enemy waves to `GameMetaData`.
+ Note that the actual generation of the `EnemySpawnUnit` is done by
+ `EnemyWaveCreator`, so this class is used to track and increase level
+ difficulty over time.
+ */
+class EndlessModeSpawnHandler {
+    var currentDifficulty: Int
+    let gameMetaData: GameMetaData
```
Comment (JermyTan) [inline GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift]: Perhaps make it `weak`?

File: GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        self.gameMetaData = gameMetaData
+        self.currentDifficulty = 0
+    }
+
+    /**
+     Increments target difficulty (until `maxDifficulty` is reached) and adds a generated
+     enemy wave corresponding to the target difficulty to `GameMetaData`.
+     */
+    func addMoreWaves() {
+        if currentDifficulty < maxDifficulty {
```
Comment (JermyTan) [inline GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift:39]: Does this mean the wave only have 5 different variants?

Comment (JermyTan) [inline GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift:39]: Does this mean the wave only have 5 different variants?

File: GameOfRunes/GameOfRunes/Scenes/GameHomeScene.swift [hunk 3]
```diff
@@ -457,11 +469,6 @@ extension GameHomeScene {
     }
     
     private func startGameSuccess() {
-        // placeholder to prevent crash
-        gameStateMachine?.avatar = .elementalWizard
-        gameStateMachine?.stage = GameViewController.storage.load(stageName: "Cathedral Mayhem")
-        
-        // @brian - room will contain everything u need for the setting up of the scenes
         gameStateMachine?.room = room
```
Comment (JermyTan) [inline GameOfRunes/GameOfRunes/Scenes/GameHomeScene.swift:497]: Should we set `avatar` and `stage` to `nil`? Likewise in single player, set `room` to `nil`?

File: GameOfRunes/GameOfRunes/GameEngine/GameMetaData.swift [hunk 4]
```diff
@@ -11,7 +11,7 @@ import Foundation
 class GameMetaData {
     // TODO: maybe change this to current avatar.
     let avatar: Avatar
-    let stage: Stage
+    var stage: Stage
```
Comment (JermyTan) [inline GameOfRunes/GameOfRunes/GameEngine/GameMetaData.swift]: Should this be immutable? Shouldnt additional waves be under `stageWaves` property?

File: GameOfRunes/GameOfRunes/GameEngine/GameEngine.swift [hunk 5]
```diff
@@ -33,7 +33,8 @@ class GameEngine: GameEngineFacade {
         metadata = GameMetaData(
             stage: stage,
             avatar: avatar,
-            manaPointsPerManaUnit: GameConfig.Mana.manaPerManaUnit
+            manaPointsPerManaUnit: GameConfig.Mana.manaPerManaUnit,
+            isEndless: stage.isEndless
```
Comment (dongsiji) [inline GameOfRunes/GameOfRunes/GameEngine/GameEngine.swift]: we don't need this right, can just add a computed property in `GameMetaData`

File: GameOfRunes/GameOfRunes/Scenes/GameHomeScene.swift (hunk 3 above)
Comment (dongsiji) [reply GameOfRunes/GameOfRunes/Scenes/GameHomeScene.swift:497]: yeah I guess that works also, multiplayer pulls from `room` while single player pulls from `avatar` and `stage`

File: GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift (hunk 2 above)
Comment (brianyenna) [reply GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift:39]: Yup!

File: GameOfRunes/GameOfRunes/Scenes/GameHomeScene.swift (hunk 3 above)
Comment (brianyenna) [reply GameOfRunes/GameOfRunes/Scenes/GameHomeScene.swift:497]: yup can. Done!

File: GameOfRunes/GameOfRunes/GameEngine/GameMetaData.swift (hunk 4 above)
Comment (brianyenna) [reply GameOfRunes/GameOfRunes/GameEngine/GameMetaData.swift]: Yup. Forgot to change back!

File: GameOfRunes/GameOfRunes/GameEngine/GameEngine.swift (hunk 5 above)
Comment (brianyenna) [reply GameOfRunes/GameOfRunes/GameEngine/GameEngine.swift]: Correct!

File: GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift (hunk 1 above)
Comment (brianyenna) [reply GameOfRunes/GameOfRunes/GameEngine/EndlessModeSpawnHandler.swift]: Done!

**Review by dongsiji — APPROVED** [review]

**Pull request merged**